File: Scala/string_algorithms/src/main/scala/burrows_wheeler/BurrowsWheeler.scala
package burrows_wheeler

object BurrowsWheeler {
  import scala.annotation.tailrec

  def calcBurrowsWheelerTransform(text: String): String = {
    val length: Int = text.length
    val doubleString: String = text * 2

    def compare(i: Int, j: Int): Boolean = {
      @tailrec
      def loop(k: Int): Boolean = {
        if (k == length) false
        else if (doubleString(i + k) < doubleString(j + k)) true
        else if (doubleString(i + k) > doubleString(j + k)) false
        else loop(k + 1)
      }
      loop(0)
    }

    val sortedIndices: List[Int] = text.indices.sortWith(compare).toList
    sortedIndices.map(ix => doubleString(ix + length - 1)).mkString
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val text: String = reader.next()
    val result: String = calcBurrowsWheelerTransform(text)
    println(result)
  }
}

File: OCaml/string_algorithms/burrows_wheeler/burrows_wheeler.ml
open Batteries

let calc_burrows_wheeler_transform text =
    let length = String.length text in
    let double_text = text ^ text in
    let compare i j =
        let rec loop k =
            if k = length then 1
            else if String.get double_text (i + k) < String.get double_text (j + k) then -1
            else if String.get double_text (i + k) > String.get double_text (j + k) then 1
            else loop (k + 1)
        in loop 0 in
    let indices = List.init length identity in
    let sorted_indices = List.sort compare indices in
    String.of_list @@ List.map (function ix -> String.get double_text (ix + length - 1)) sorted_indices


let () =
    let text = read_line() in
    let result = calc_burrows_wheeler_transform text in
    print_endline result
